[Mac] Allow audio device selection

hiroshihorie opened the pull request:
### Goals

- [x]  Audio device list & selection at runtime.
- [x]  Detect and switch device when default device changes at runtime. (Connect AirPods etc)
- [ ] Feed audio sample buffers directly.


`bypassAudioProcessing` is _temporarily removed_, but I will make it available again.
-> Now available again.

hiroshihorie committed:
first attempt

hiroshihorie committed:
remove unused dep

hiroshihorie committed:
init playout / recording

hiroshihorie committed:
use AudioDeviceID as guid

hiroshihorie committed:
switch device method

hiroshihorie committed:
equality

hiroshihorie committed:
default device

hiroshihorie committed:
`isDefault` property

hiroshihorie committed:
dont format default device name

hiroshihorie committed:
type param

hiroshihorie committed:
bypass

hiroshihorie committed:
refactor

davidzhao commented on modules/audio_device/mac/audio_device_mac.cc:
```diff
@@ -867,12 +864,9 @@ int32_t AudioDeviceMac::RecordingDeviceName(uint16_t index,
 
   memset(name, 0, kAdmMaxDeviceNameSize);
-
-  if (guid != NULL) {
```
can anything else call this function and pass in NULL for `RecordingDeviceName` ?

hiroshihorie replied:
`RecordingDeviceName` is supposed to populate `name` and `guid` of the device,
but populating `guid` was not implemented in their code.

So I implemented to return the `AudioDeviceID` which is not technically `guid` but 
it works for identifying the device.

davidzhao reviewed (approved):
incredible work @hiroshihorie!

For the new files such as: RTCAudioDeviceModule.mm, are those from somewhere else in the webrtc tree?

hiroshihorie commented:
Yes there is a `webrtc::AudioDeviceModule` class.
https://chromium.googlesource.com/external/webrtc/stable/webrtc/+/15a8a6e7edb2c8df7f5b21e759ace213a4ab7292/modules/audio_device/include/audio_device.h#19

hiroshihorie committed:
fix

hiroshihorie committed:
append Audio to thread labels

hiroshihorie committed:
ref

davidzhao commented:
> Yes there is a `webrtc::AudioDeviceModule` class. https://chromium.googlesource.com/external/webrtc/stable/webrtc/+/15a8a6e7edb2c8df7f5b21e759ace213a4ab7292/modules/audio_device/include/audio_device.h#19

But is the ObjC wrapper class from the tree? or is that a new file you authored?

hiroshihorie commented:
Yes there's a few new files I authored.

for `RTCAudioDeviceModule`, In the future,
I might want to add ability to create virtual devices.
`RTCAudioDeviceModule.addVirtualDevice(...)`

then user can call the existing (implemented in this PR) `RTCAudioDeviceModule.setInputDevice(...)`,
to switch to their virtual device.

This virtual device class will conform to a new `RTCVirtualAudioDevice` protocol (or base class) so  users can feed their own `CMSampleBuffer`s through here.

This is just a thought for design 🙂

I want to achieve unified design that will work for both iOS / Mac platforms, but the underlying audio system is pretty different for each platform. So I'm not sure if it's possible.

Another concern is mixing the audio, for example game streaming, user would like to speak while having game audio. `ReplayKit` delivers audio buffer separately for `app` and `mic`.
https://developer.apple.com/documentation/replaykit/rpsamplebuffertype
So it needs to get mixed somehow... still investigating ...

davidzhao commented:
Got it, for the new files, could you add a LiveKit copyright in it? This will help differentiate which files originated from upstream vs us.

davidzhao commented:
> This virtual device class will conform to a new `RTCVirtualAudioDevice` protocol (or base class) so users can feed their own `CMSampleBuffer`s through here.
> 
> This is just a thought for design 🙂
> 
> I want to achieve unified design that will work for both iOS / Mac platforms, but the underlying audio system is pretty different for each platform. So I'm not sure if it's possible.

YES! This is going to be awesome!!

hiroshihorie committed:
lk headers

hiroshihorie committed:
low level apis

hiroshihorie committed:
Merge branch 'm97_release' into expose-audio-device-module

hiroshihorie committed:
fix thread checks

Some methods of ADM needs to be run on worker thread, otherwise RTC's thread check will fail.

hiroshihorie commented:
RTC's thread checks were failing so improved thread handling.
All access to ADM methods are from worker thread now.

`RTC_DCHECK_RUN_ON(&main_thread_checker_);` checks whether its called from the same thread that the ADM was created on (which must be created on worker thread).
https://github.com/webrtc-sdk/webrtc/blob/main/modules/audio_device/audio_device_buffer.cc#L135
https://github.com/webrtc-sdk/webrtc/blob/main/modules/audio_device/audio_device_buffer.cc#L149

hiroshihorie commented:
## Additional required logic

### Active device (lost) disconnected ✅
Current: Nothing happens (audio input/output stops)
Expected: Switch to `default` device and continue input/output.

### OS's default device changed ✅
Current: Audio input/output continues on the same device even if `default` device is changed.
Expected: if `default` device is switched by the OS, change input/output to the new `default` device.

hiroshihorie committed:
switch to default device when removed

hiroshihorie committed:
close mixerManager if didn't switch to default device

hiroshihorie committed:
default audio device switched

hiroshihorie commented:
Above 2 logic have been implemented and seems to work as expected.

hiroshihorie committed:
expose devices update handler

hiroshihorie commented:
implemented `audioDeviceModule.setDevicesUpdatedHandler` to get notification when
audio input / output devices update ( or default changes )

cloudwebrtc commented:
I think we need to compile it into an `xframework` to make sure the API works correctly on macOS and doesn't break iOS

hiroshihorie committed:
fix ios compile

hiroshihorie commented:
I made some changes and iOS compiles

hiroshihorie committed:
fix bug: don't always recreate RTCAudioDeviceModule

hiroshihorie commented:
Fixed bug where `AudioDeviceSink` gets reset since `RTCAudioDeviceModule` was being recreated on every access.

cloudwebrtc commented:
hey @hiroshihorie, It seems that changing the PR from target m97 to m104 introduces additional diffs, should we still target m97 after merging we get the minimal diff for the patch?

hiroshihorie commented:
> hey @hiroshihorie, It seems that changing the PR from target m97 to m104 introduces additional diffs, should we still target m97 after merging we get the minimal diff for the patch?

Yup, I changed the target which wasn't a good idea..

cloudwebrtc commented:
The code looks good.

davidzhao commented:
@hiroshihorie shall we merge this as well as into 104?
